digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala

```diff
@@ -114,26 +125,30 @@ object ProjectDeltaCalculator {
     * @return Map between the sections old -> new
     */
   def partition(roadAddresses: Seq[RoadAddress], projectLinks: Seq[ProjectLink]): Map[RoadAddressSection, RoadAddressSection] = {
-    val groupedAddresses = roadAddresses.sortBy(_.startAddrMValue).groupBy(ra => (ra.roadNumber, ra.roadPartNumber, ra.track))
-    val addressesGroups = groupedAddresses.mapValues(v => combine(v.sortBy(_.startAddrMValue))).values.flatten.map(ra =>
-      RoadAddressSection(ra.roadNumber, ra.roadPartNumber, ra.roadPartNumber,
-        ra.track, ra.startAddrMValue, ra.endAddrMValue, ra.discontinuity, ra.roadType)
-    ).toSeq
-
-    val groupedProjectLinks = projectLinks.sortBy(_.startAddrMValue).groupBy(pl => (pl.roadNumber, pl.roadPartNumber, pl.track))
-    val projectLinksGroups = groupedProjectLinks.mapValues(v => combine(v.sortBy(_.startAddrMValue))).values.flatten.map(pl =>
-      RoadAddressSection(pl.roadNumber, pl.roadPartNumber, pl.roadPartNumber,
-        pl.track, pl.startAddrMValue, pl.endAddrMValue, pl.discontinuity, pl.roadType)
-    ).toSeq
 
-    addressesGroups.map { sec =>
-      val linkId = roadAddresses.find(ra => sec.includes(ra)).map(_.linkId).get
-      val targetGroup = projectLinksGroups.find(_.includes(projectLinks.find(_.linkId == linkId).get))
-      sec -> targetGroup.get
+    val paired = pair(roadAddresses, projectLinks).groupBy(x => (x._1.roadNumber, x._1.roadPartNumber, x._2.roadNumber, x._2.roadPartNumber))
+
+    val (oldL, newL) = (paired.map( x=> x._2.map(_._1)),paired.map( x=> x._2.map(_._2)))
+
+    val sourceCombined = oldL.map(o => {
+      o.sortBy(_.startAddrMValue).groupBy(ra => (ra.roadNumber, ra.roadPartNumber, ra.track))
+        .mapValues(v => combine(v.sortBy(_.startAddrMValue))).values.flatten.map(ra =>
+        RoadAddressSection(ra.roadNumber, ra.roadPartNumber, ra.roadPartNumber,
+          ra.track, ra.startAddrMValue, ra.endAddrMValue, ra.discontinuity, ra.roadType)).toSeq
+    })
+    val targetCombined = newL.map(n => {
+        n.sortBy(_.startAddrMValue).groupBy(ra => (ra.roadNumber, ra.roadPartNumber, ra.track))
+          .mapValues(v => combine(v.sortBy(_.startAddrMValue))).values.flatten.map(ra =>
+          RoadAddressSection(ra.roadNumber, ra.roadPartNumber, ra.roadPartNumber,
+            ra.track, ra.startAddrMValue, ra.endAddrMValue, ra.discontinuity, ra.roadType)).toSeq
+      })
+
+    sourceCombined.map { sec =>
+      val linkId = roadAddresses.find(ra => sec.head.includes(ra)).map(_.linkId).get
+      val targetGroup = targetCombined.find(tc => tc.head.includes(projectLinks.find(_.linkId == linkId).get))
+      sec.head -> targetGroup.get.head
```

Comment:
why sec.head? What happens to other elements?

Reply:
Because that list always contain one element with a seq, anyway the list is now flatten so there is not need to .head